### PR TITLE
fix(agent): persist direct-stream replay checkpoints

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

- carry the trusted runtime message identity and server-resolved replay checkpoints through the signed control-plane request
- durably append each private provider replay checkpoint to the exact run with the opaque run-writer credential, and await the API acknowledgement before execution can continue
- keep checkpoint payloads off the runtime SSE stream while retaining a private turn-complete boundary so the paired API stream cannot pass tool or terminal output early
- fail closed on provider-stream errors, missing message identity, missing writer authority, append failure, timeout, or run cancellation
- keep the public control-plane request type source-compatible by leaving `messageId` optional

## Deployment order

Merge and deploy veryfront/veryfront-api#4545 first. The run-scoped append route from veryfront/veryfront-api#4537 is the durable checkpoint sink. Runtime emission remains disabled unless the host sets `VERYFRONT_ENABLE_PROVIDER_REPLAY_CHECKPOINT_EMISSION=1` or the run is continuing an existing checkpoint.

Closes part of veryfront/veryfront-issue-inbox#522.

## Verification

- focused suite: 7 files, 197 checks
- `deno task lint`
- `deno task typecheck`
- `deno task lint:test-typecheck`
- `deno task test:layout`
- `deno task lint:test-semantic-dispositions`
- `deno task docs:api-reference:check`
- `git diff --check`
- full unit suite attempted locally; the host has no Node/npm, so the unrelated CLI dependency-install test stops at `Failed to spawn npm: entity not found`; exact-head CI provides the complete unit result

## Summary by CodeRabbit

- **New Features**
  - Agent execution streams can emit provider replay checkpoint events for improved continuity and replay handling.
  - Runtime agent requests preserve message identity and replay checkpoint data throughout processing.
  - Checkpoints are persisted with acknowledged completion and support cancellation and timeout handling.

- **Bug Fixes**
  - Improved validation and failure reporting for checkpoint persistence and missing runtime identity.

- **Documentation**
  - Updated API reference links to current source locations.

### CODE REVIEW

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
## Automated Code Review — Score: 81/100 (Good, minor suggestions)

Solid, well-scoped change that threads a `messageId` through the internal control-plane stream so provider replay checkpoints can be persisted and resumed correctly. Follows existing codebase idioms closely and the trust model is sound.

**Strengths**
- `messageId` is plumbed consistently through the full path (`agent-invocation-contract.ts` → `schema.ts`/`toRuntimeRunAgentInput` → `run-stream.ts`), matching the existing optional-field pattern (`taskId`, `parentRunId`) rather than inventing a new convention.
- The fail-closed check (`providerReplayCheckpointEmissionEnabled === true && !input.messageId` → throw) is correctly scoped: it doesn't block the "continue an existing checkpoint" branch, which can't fire without a matching `messageId` anyway.
- `agent-stream.handler.ts` hardcodes `serverEnvelopeVerified: true` when calling `getServerResolvedProviderReplayCheckpoints`, but that's consistent with the rest of the handler — the whole request is already authenticated via `verifyControlPlaneRequest`/`verifyControlPlaneRequest` at entry (see the existing "veryfront-api is the trusted control-plane caller" comment nearby), so this isn't a bypass.
- Snapshotting `isProviderReplayCheckpointEmissionEnabled()` once into `defaultDeps` (module load, before any `runWithProjectEnv` isolation) correctly captures the *host* gate rather than a project-isolated env — matches the PR description's stated intent and prevents project code from spoofing the flag.
- New tests cover the two most important paths: successful emission and the fail-closed no-`messageId` case.

**Concerns**
1. **Untested edge case**: `providerReplayCheckpointRelay.attach()`'s callback calls `controller.enqueue(...)` directly and throws `"...detached before persistence"` if `!clientAttached`, unlike every other relay in this file (`enqueueIfAttached`, `enqueueHeartbeatIfAttached`), which wrap `enqueue` in try/catch and silently flip `clientAttached = false` instead. This divergence looks like a deliberate "don't silently lose a checkpoint" choice, but it's unverified: please confirm a rejected `__vfPersistProviderReplayCheckpoint` actually propagates as a clean run failure through the runtime rather than becoming an unhandled rejection.
2. **Missing test**: the second half of `shouldEmitProviderReplayCheckpoints` — continuing an existing checkpoint whose `messageId` matches `input.messageId` while `providerReplayCheckpointEmissionEnabled` is falsy — isn't covered by any new test, even though it's called out explicitly in the PR description ("forward server-resolved checkpoints into the direct runtime").
3. No handler-level test (`agent-stream.handler.test.ts`) exercises the new `providerReplayCheckpoints` wiring end-to-end (e.g., a malformed `serverResolvedProviderReplayCheckpoints` in `forwardedProps` reaching `assertReconstructibleProviderReplayCheckpoint` and failing the request) — the PR only tests at the lower `run-stream.ts` level with directly-injected deps.
4. Minor: `createProviderReplayCheckpointRelay` duplicates `createModelCallContextRelay`'s buffer/attach structure almost verbatim; could be generalized into one shared helper, though not blocking.

**Suggested action before merge**: add the resume-continuation test (concern #2) since it's the one explicitly promised by the PR description and currently unverified, and get a quick confirmation on the unhandled-rejection question (#1) from whoever owns the runtime package. Neither blocks merge on its own, but together they're the difference between "good" and "excellent" here.

Note: PR is `mergeable_state: blocked` and CI (deno/bun/node test matrix) was still running at review time — recommend confirming green CI before merge, independent of this review's score.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/internal-agents/run-stream.ts [hunk 1]
```diff
@@ -1151,6 +1200,16 @@ export async function createRuntimeAgentStreamResponse(
...
+            providerReplayCheckpointRelay.attach((event) => {
+              if (!clientAttached) {
+                throw new Error(
+                  "Provider replay checkpoint stream detached before persistence",
+                );
+              }
+              controller.enqueue(
+                formatAgUiEvent(PROVIDER_REPLAY_CHECKPOINT_SSE_EVENT_NAME, event),
+              );
+            });
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts]: Addressed on exact head 7e1d2da5e. The private enqueue now marks clientAttached false before throwing on controller failure. Rejected checkpoint persisters are awaited by the runtime and stop execution; provider-replay-emission.test.ts now pins that behavior with a rejecting sink and verifies the model runs only once. Focused tests, lint, typecheck, and diff checks pass.

File: src/internal-agents/run-stream.ts [hunk 2]
```diff
@@ -953,6 +982,17 @@ export async function createRuntimeAgentStreamResponse(
       });
     };
     const systemPrompt = await resolveSystemPrompt();
+    const providerReplayCheckpoints = deps.providerReplayCheckpoints ?? [];
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts:1037]: Addressed on exact head 7e1d2da5e. Added a resume regression with the host gate disabled and a matching server-resolved checkpoint. It verifies that persistence is wired and the private checkpoint frame is emitted. Focused tests, lint, typecheck, and diff checks pass.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review

File: src/server/handlers/request/agent-stream.handler.ts [hunk 3]
```diff
@@ -1151,6 +1154,11 @@ export class AgentStreamHandler extends BaseHandler {
                           toRuntimeRunAgentInput(payload),
                           runtimeBaseAgent as Agent,
                         );
+                        const providerReplayCheckpoints =
+                          getServerResolvedProviderReplayCheckpoints(
+                            runtimeInput.forwardedProps,
+                            true,
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/handlers/request/agent-stream.handler.ts]: Fixed on exact head 399a9dbda. The strict internal schema and canonical runtime input now retain the signed top-level serverResolvedProviderReplayCheckpoints field. AgentStreamHandler passes the field through the object overload only when present, preserving the forwardedProps fallback when absent. A schema regression pins the top-level handoff; 164 focused checks, test-typecheck, lint, typecheck, and docs checks pass.

**Comment by kojiwakayama** [comment]
@codex review

**Review by kwakayama — CHANGES_REQUESTED** [review]
Blocking findings on exact head `399a9dbda5`:

1. `createProviderReplayCheckpointRelay.persist` resolves after buffering an SSE frame, not after the API durably appends it. The runtime treats this callback as the persistence barrier, so a local tool or the next provider request can begin before the checkpoint is stored. A disconnect or append failure can therefore leave irreversible continuation effects without replay state. This needs an acknowledged durable append before the runtime continuation is released.

2. If provider execution fails after `StepStarted` but before the persistence wrapper runs, neither `complete` nor `fail` settles the open relay turn. The EOF flush waits forever in `takeCompletedTurn`, preventing terminal delivery and cleanup. The runtime error and abort paths must reject any pending turn.

3. `RuntimeAgentControlPlaneStreamRequest.messageId` changed from absent to required even though this is an exported public payload type and the receiver intentionally remains backward-compatible. Keep it optional while having the builder populate it for new invocations.

Please add regressions that prove: continuation cannot start before acknowledged persistence, provider failure closes the stream and cleanup runs, and the legacy request shape still type-checks.

File: src/internal-agents/run-stream.ts [hunk 4]
```diff
@@ -1195,12 +1350,17 @@ export async function createRuntimeAgentStreamResponse(
...
             for (const event of trailingEvents.events) {
               for (const mappedEvent of mapRuntimeEventToAgUi(state, event)) {
-                prepareToolResultIfNeeded(mappedEvent.event, mappedEvent.payload);
-                enqueueIfAttached(mappedEvent.event, mappedEvent.payload);
+                await emitMappedEvent(mappedEvent);
               }
             }
 
             throwIfAborted();
+            await flushProviderReplayTurn();
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts:1362]: Fixed on exact head 6b27ebf5c. An open replay turn is now failed when the mapped runtime emits RunError, so EOF cleanup never waits on an unset boundary. A provider-error regression emits StepStarted followed by RunError with emission enabled and verifies the response closes as failed.

File: src/agent/runtime/agent-invocation-contract.ts [hunk 5]
```diff
@@ -430,6 +430,7 @@ export type RuntimeAgentControlPlaneStreamRequest = {
   agentId: RuntimeAgentRunContext["agentId"];
   threadId: RuntimeAgentRunContext["conversationId"];
   runId: RuntimeAgentRunContext["runId"];
+  messageId: RuntimeAgentRunContext["messageId"];
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/agent-invocation-contract.ts]: Fixed on exact head 6b27ebf5c. RuntimeAgentControlPlaneStreamRequest.messageId is optional again for source compatibility, while the canonical builder continues to populate it and gated emission fails closed if the receiving runtime has no message identity. Generated docs are current.

File: src/internal-agents/run-stream.ts [hunk 6]
```diff
@@ -779,6 +789,84 @@ function createModelCallContextRelay(
...
+    resolve(frames);
+  };
+
+  return {
+    persist: async (checkpoint) => {
+      buffered.push({
+        event: PROVIDER_REPLAY_CHECKPOINT_SSE_EVENT_NAME,
+        payload: createProviderReplayCheckpointEvent(checkpoint),
+      });
+      resolveIfReady();
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts]: Fixed on exact head 6b27ebf5c. Direct streams no longer treat SSE buffering as durability. AgentStreamHandler binds the existing signed X-Veryfront-Run-Event-Token to a run-scoped API persister, and the runtime awaits POST /runs/{run_id}/events successfully before completing the provider turn. The SSE carries only a safe turn boundary, never opaque blocks. Focused tests cover the exact request, awaited success, redacted failure, and ordering before ToolCallEnd.

File: src/internal-agents/run-stream.ts [hunk 7]
```diff
@@ -1151,6 +1268,45 @@ export async function createRuntimeAgentStreamResponse(
...
+                clientAttached = false;
+                throw new Error(
+                  "Provider replay checkpoint stream detached before persistence",
+                );
+              }
+            };
+            let providerReplayStepOpen = false;
+            const flushProviderReplayTurn = async () => {
+              if (!providerReplayStepOpen) return;
+              for (const frame of await providerReplayCheckpointRelay.takeCompletedTurn()) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts:1291]: Fixed on exact head 6b27ebf5c. RunError settles the open replay relay before EOF finalization, and provider checkpoint hook failures reject the waiter through the private failure hook. Focused tests cover both provider-stream failure and persister rejection without hanging.

File: src/internal-agents/run-stream.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/internal-agents/run-stream.ts]: Fixed on exact head 6b27ebf5c. The persistence hook now awaits the run-scoped API append with the writer token already minted on control-plane requests. Only after the API responds successfully does the runtime emit the safe turn boundary and release ToolCallEnd or terminal output.

File: src/agent/runtime/agent-invocation-contract.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/agent-invocation-contract.ts]: Fixed on exact head 6b27ebf5c. The exported messageId property is optional for compatibility; buildRuntimeAgentControlPlaneStreamRequestFromInvocation still sets it for canonical invocations.

**Comment by kojiwakayama** [comment]
@codex review

File: src/agent/runtime/provider-replay-emission.test.ts [hunk 8]
```diff
@@ -185,6 +188,7 @@ describe("provider replay checkpoint emission", () => {
       }
 
       assertEquals(operations.indexOf("persist:done") < operations.indexOf("model:2"), true);
+      assertEquals(operations.indexOf("turn:complete") < operations.indexOf("model:2"), true);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/provider-replay-emission.test.ts]: Fixed in ff13f8f9dce229a4f02914a71c98b691dc2e96be: the test now first asserts that turn:complete exists, then checks that its index precedes the second model invocation.

File: src/internal-agents/run-stream.ts [hunk 9]
```diff
@@ -1151,6 +1264,49 @@ export async function createRuntimeAgentStreamResponse(
...
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/internal-agents/run-stream.ts:1291]: Fixed in ff13f8f9dce229a4f02914a71c98b691dc2e96be: run cancellation now fails the pending replay relay before cancelling the runtime reader. The new cancellation regression verifies response completion, reader cancellation, and session cleanup.

File: src/internal-agents/provider-replay-checkpoint-persister.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+/** Create an exact-run checkpoint writer backed by the API's durable append route. */
+export function createRunScopedProviderReplayCheckpointPersister(input: {
+  apiUrl: string;
+  runId: string;
+  runEventToken: string;
+  fetch?: typeof globalThis.fetch;
+}): (checkpoint: ProviderReplayCheckpoint) => Promise<void> {
+  const fetchImpl = input.fetch ?? createOriginBoundOutboundFetch(input.apiUrl);
+  const url = new URL(`/runs/${encodeURIComponent(input.runId)}/events`, input.apiUrl);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/internal-agents/provider-replay-checkpoint-persister.ts]: Fixed in 663c489d3e79258619209fec62468c36eee5f816 and retained in exact head ff13f8f9dce229a4f02914a71c98b691dc2e96be: the canonical endpoint resolver preserves an API base pathname such as /api, with a regression asserting the exact append URL.

File: src/internal-agents/run-stream.ts [hunk 11]
```diff
@@ -1151,6 +1264,49 @@ export async function createRuntimeAgentStreamResponse(
...
+            const emitMappedEvent = async (mappedEvent: {
+              event: string;
+              payload: Record<string, unknown>;
+            }) => {
+              if (mappedEvent.event === "StepStarted") {
+                await flushProviderReplayTurn();
+                providerReplayStepOpen = shouldEmitProviderReplayCheckpoints;
+              }
+              if (mappedEvent.event === "ToolCallEnd") {
+                await flushProviderReplayTurn();
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/internal-agents/run-stream.ts:1305]: Fixed in ff13f8f9dce229a4f02914a71c98b691dc2e96be: each runtime call now owns an idempotent failure signal invoked by the stream error path independently of SSE consumption. A new step/tool/error-without-completion regression proves a pending ToolCallEnd boundary is released and the response closes with RunError.

File: src/internal-agents/provider-replay-checkpoint-persister.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+const jsonStringify = JSON.stringify;
+const textEncoder = new TextEncoder();
+
+function snapshotFetch(fetchImpl: Fetch): Fetch {
+  return (input, init) => apply(fetchImpl, undefined, [input, init]) as Promise<Response>;
+}
+
+function isValidRunEventAppendToken(token: string | null | undefined): token is string {
+  return typeof token === "string" && token.length > 0 && token.trim() === token &&
+    textEncoder.encode(token).byteLength <= MAX_RUN_EVENT_APPEND_TOKEN_BYTES;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/internal-agents/provider-replay-checkpoint-persister.ts]: Fixed in 44551cb3a4e8b164115ec3c38c6a38d7e73b477c: credential validation now uses module-captured trim, TextEncoder encode, and typed-array byte-length intrinsics. The new shared-realm poisoning regression verifies that the opaque writer token never reaches poisoned String or TextEncoder prototype methods.

File: src/internal-agents/provider-replay-checkpoint-persister.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    const timeout = setTimeout(() => controller.abort(timeoutError), timeoutMs);
+    const onAbort = () => controller.abort(getAbortReason(abortSignal!));
+    abortSignal?.addEventListener("abort", onAbort, { once: true });
+
+    try {
+      const response = await fetchImpl(url, {
+        method: "POST",
+        headers: {
+          Accept: "application/json",
+          Authorization: `Bearer ${token}`,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/internal-agents/provider-replay-checkpoint-persister.ts:102]: No source change: this is an intentional host-configuration boundary, not a project-controlled URL. The handler resolves apiUrl from host env after project discovery and overwrites project VERYFRONT_API_URL values. The shared resolver contract deliberately supports operator-owned HTTP service endpoints (covered by the http://veryfront-api.veryfront-staging.svc resolver regression), while the persister uses the DNS-pinned, same-origin outbound boundary and blocks redirects or origin changes. A blanket HTTPS requirement here would break the supported in-cluster API path and would be inconsistent with the existing run-writer and platform-client policy; changing that transport policy belongs at the shared host resolver boundary.

**Comment by kojiwakayama** [comment]
@kwakayama Exact head `861ea484a144cd117aeadf7d7b32aae416550b60` now passes the focused credential tests, semantic unit-boundary audit, lint, typecheck, and automated review. All review threads are resolved. Please replace the stale change request with exact-head approval when ready; auto-merge is already enabled.

**Review by kwakayama — APPROVED** [review]
Reviewed exact head 861ea484a144cd117aeadf7d7b32aae416550b60. All actionable findings are addressed, the remaining transport-policy thread was resolved with repository-contract evidence, and automated review is green with zero unresolved threads.
